### PR TITLE
[SPARK-30198][Core] BytesToBytesMap does not grow internal long array as expected

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch changes the condition to check if BytesToBytesMap should grow up its internal array. Specifically, it changes to compare by the capacity of the array, instead of its size.

### Why are the changes needed?

One Spark job on our cluster hangs forever at BytesToBytesMap.safeLookup. After inspecting, the long array size is 536870912.

Currently in BytesToBytesMap.append, we only grow the internal array if the size of the array is less than its MAX_CAPACITY that is 536870912. So in above case, the array can not be grown up, and safeLookup can not find an empty slot forever.

But it is wrong because we use two array entries per key, so the array size is twice the capacity. We should compare the current capacity of the array, instead of its size.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

This issue only happens when loading big number of values into BytesToBytesMap, so it is hard to do unit test. This is tested manually with internal Spark job.

This is a manual test case. Once the number of keys reaches 268435456, lookup will hang forever.

```java
@Test
  public void testCapacity() {
    TestMemoryManager memoryManager2 =
            new TestMemoryManager(
                    new SparkConf()
                            .set(package$.MODULE$.MEMORY_OFFHEAP_ENABLED(), true)
                            .set(package$.MODULE$.MEMORY_OFFHEAP_SIZE(), 25600 * 1024 * 1024L)
                            .set(package$.MODULE$.SHUFFLE_SPILL_COMPRESS(), false)
                            .set(package$.MODULE$.SHUFFLE_COMPRESS(), false));
    TaskMemoryManager taskMemoryManager2 = new TaskMemoryManager(memoryManager2, 0);
    final long pageSizeBytes = 8000000 + 8; // 8 bytes for end-of-page marker
    final BytesToBytesMap map = new BytesToBytesMap(taskMemoryManager2, 1024, pageSizeBytes);

    try {
      for (long i = 0; i < BytesToBytesMap.MAX_CAPACITY; i++) {
        final long[] value = new long[]{i};
        map.lookup(value, Platform.LONG_ARRAY_OFFSET, 8).append(
                value,
                Platform.LONG_ARRAY_OFFSET,
                8,
                value,
                Platform.LONG_ARRAY_OFFSET,
                8);
      }
      map.free();
    } finally {
      map.free();
    }
```

### CODE REVIEW

**Comment by viirya** [comment]
retest this please.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (maropu) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Probably, we need to leave some comments about why the size needs to be divided by 2.

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: thanks! Added few comments.

**Comment by viirya** [comment]
cc @cloud-fan

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (viirya) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Actually, I also think that we should false canGrowArray like: 

```scala
if (numKeys >= growthThreshold && longArray.size() / 2 >= MAX_CAPACITY) {
  canGrowArray = false;
}
```

So as we reach max capacity of the map, canGrowArray is set to false. We can fail next append and let the map spill and fallback to sort-based aggregation in HashAggregate. Thus we can prevent a similar forever-loop happens when we reach max capacity. 

cc @cloud-fan @felixcheung

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (maropu) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Thanks! It looks good to me.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (cloud-fan) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: can you remind me some more details of `BytesToBytesMap`? What happens if we don't grow? I don't see a loop in this method and not sure how the job hangs.

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Ah, sure.

The client of `BytesToBytesMap`, like HashAggregate, will call `lookup` to find a `Location` to write value. The returned location will be used to do append (`Location.append`).  Everytime after we append a key/value, we check if it is time to grow internal array and grow up it if needed.

`lookup` delegates looking up keys to `safeLookup`. Its control flow looks like:

```java
int pos = hash & mask;
int step = 1;
// an infinite loop until find matching key or empty slot.
while (true) {
  if (empty slot found) {
    ...
  } else {
    // check if matching key
    ...
  }
  // increase pos and step
  pos = (pos + step) & mask;
  step++;
}
```

So the job hangs in this loop because it can not find any empty location as the internal array is full.

We early stop growing the internal array due to wrongly check array size at:

```java
if (numKeys >= growthThreshold && longArray.size() < MAX_CAPACITY) {
  ...
}
```

Another point https://github.com/apache/spark/pull/26828#discussion_r356370350 is we may want to set `canGrowArray` to false once we are close to max capacity, so we can avoid infinite loop again.

Comment (cloud-fan) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: shouldn't `lookup` throw OOM if no space can be found?

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (dongjoon-hyun) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: @viirya . I agree your issue, but `longArray.size() / 2` might be too strict because we cap with `MAX_CAPACITY`. In other words, there is a room to grow still.
```
 // Allocate the new data structures
    allocate(Math.min(growthStrategy.nextCapacity(oldCapacity), MAX_CAPACITY));
```

Can we have explicit test cases for these boundary conditions?

**Comment by dongjoon-hyun** [comment]
cc @dbtsai

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: `lookup` just looks up empty slot in the internal array for a new key. It does not allocate memory. The array is allocated/grown up in last time `append`.

Once an empty slot (a `Location` object) is found, the client of `BytesToBytesMap` may call `append` to the Location, OOM could be thrown during append key/value.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: The max capacity is big number. Is it ok to have unit test allocating such big array?

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (cloud-fan) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: let me confirm the problem: so `append` mistakenly think there is enough space, and doesn't grow the array. This makes the client of `BytesToBytesMap` keeping calling `lookup` and hand. Is my understanding correct?

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: `append` mistakenly think it can not grow the array anymore so does not grow the array. It keeps append value until full. Then the client calling `lookup` can not find an empty slot and gets stuck in infinite loop.

Comment (cloud-fan) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: I see, the `append` doesn't grow the array while it should. This makes `BytesToBytesMap` malformed (the array is not big enough to serve the data region) and causes problems.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (dongjoon-hyun) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: I guessed we can use `mock` with simple `growAndRehash` (although I didn't try).

Comment (dongjoon-hyun) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Never mind. I don't want to block this PR and you. I'll try that later by myself.

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Ok. Sounds good. I will also do test to see if I can add it. Thanks for the suggestion!

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: hmm, even mock with `growAndRehash` to avoid allocate the array, `append` still needs to allocate memory page for values to insert. Mock with `append` makes less sense because it is where this logic remains.

Comment (dongjoon-hyun) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Oh, you already tested it. Got it. Thank you for spending time for that.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master/2.4.

Since `MAX_CAPACITY` is defined as `The maximum number of keys that BytesToBytesMap supports` and `longArray` is defined as `A single array to store the key and value`, this PR is correct.

**Comment by dongjoon-hyun** [comment]
cc @JoshRosen , @gatorsmile , @wangyum (since he is a release manager for 3.0.0-preview2).

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (cloud-fan) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: If we do this, we won't call `growAndRehash` here, is it expected?

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Oh, I was not meaning to replace current condition, but to add anoother check.

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: let me think about it. if making sense, will submit another PR for it.

**Comment by cloud-fan** [comment]
Can we at least provide a manual regression test in the PR description?  so that people can try and evaluate the risk.

**Comment by viirya** [comment]
> Can we at least provide a manual regression test in the PR description? so that people can try and evaluate the risk.

Good suggestion. I added one manual test case in the PR description.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java
Comment (ankurdave) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:748]: @viirya I'm encountering the same problem that you describe here. When the map is close to `MAX_CAPACITY` and needs to grow, `numKeys >= growthThreshold && longArray.size() / 2 >= MAX_CAPACITY` is true. This prevents the map from resizing, but currently `canGrowArray` remains `true`. Therefore the map keeps accepting new keys and exceeds its growth threshold. This ultimately causes the query to fail in the UnsafeKVExternalSorter constructor.

It looks like you didn't submit a PR for this - is there a reason why not? If there's no problem with your suggested fix, I can submit a PR now.

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:748]: I submitted another PR targeting the case of reaching `MAX_CAPACITY`. It actually prevents we increase the number of keys to be `MAX_CAPACITY`.

Comment (ankurdave) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:748]: Thanks for the quick response! I saw that PR (https://github.com/apache/spark/pull/26914) but I don't think it solves the problem I'm encountering. That PR stops accepting new keys once we have reached `MAX_CAPACITY - 1` keys, but this is too late. By that time, we will have far exceeded the growth threshold. If we attempt to spill the map in this state, the UnsafeKVExternalSorter will not be able to reuse the long array for sorting, causing the query to fail.

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:748]: I think the problem I posted above, is when we reach `MAX_CAPACITY`, a forever-loop happens during calling lookup. The previous PR fixed it. Sounds like you are encountering another problem?

Comment (viirya) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:748]: >  If we attempt to spill the map in this state, the UnsafeKVExternalSorter will not be able to reuse the long array for sorting, causing the query to fail.

In `UnsafeKVExternalSorter`, it will check if the long array can be reused or not. Isn't? If it cannot be reused, a new pointer array will be created, no?

Comment (ankurdave) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:748]: > Sounds like you are encountering another problem?

You're right, it's not the same problem.

> In UnsafeKVExternalSorter, it will check if the long array can be reused or not. Isn't? If it cannot be reused, a new pointer array will be created, no?

Yes, but by this point the task has typically consumed all available memory, so the allocation of the new pointer array is likely to fail.

Comment (ankurdave) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:748]: I filed [SPARK-32872](https://issues.apache.org/jira/browse/SPARK-32872) and submitted https://github.com/apache/spark/pull/29744 to fix this.

**Pull request closed without merging**
